netlify-functions/src/main/scala/feral/netlify/IOLambda.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.netlify
+
+import cats.effect.IO
+import cats.effect.IOLocal
+import cats.effect.kernel.Resource
+import feral.IOSetup
+import io.circe.Decoder
+import io.circe.Encoder
+import io.circe.scalajs._
+
+import scala.scalajs.js
+import scala.scalajs.js.JSConverters._
+import scala.scalajs.js.|
+
+abstract class IOLambda[Event, Result](
+    implicit private[netlify] val decoder: Decoder[Event],
+    private[netlify] val encoder: Encoder[Result]
+) extends IOSetup {
+
+  final type Setup = (Event, Context[IO]) => IO[Option[Result]]
+
+  final override protected def setup: Resource[IO, Setup] =
+    handler.map { handler => (event, context) =>
+      for {
+        event <- IOLocal(event)
+        context <- IOLocal(context)
+        env = LambdaEnv.ioLambdaEnv(event, context)
+        result <- handler(env)
+      } yield result
+    }
+
+  def main(args: Array[String]): Unit = {
+    val handlerName = getClass.getSimpleName.init
+    js.Dynamic.global.exports.updateDynamic(handlerName)(handlerFn)
+  }
```

Comment:
So IIUC _all_ netlify serverless functions have to export a method named `handler`. So I wonder if we can just simplify the design here and use an idiomatic `@JSExport`.

Reply:
I don't see why not, that makes the most sense.

Reply:
guess we can't do a top level export in an abstract class... I've changed this to use a constant string but I don't know if we can do much more

Reply:
Oh sorry, I missed this comment before.